File: src/models/dm.cpp
#include <microscopes/models/dm.hpp>
#include <distributions/special.hpp>

using namespace std;
using namespace distributions;
using namespace microscopes::common;
using namespace microscopes::models;

void
dm_group::add_value(const hypers &m, const value_accessor &value, rng_t &rng)
{
  MICROSCOPES_ASSERT(value.shape() == categories());
  unsigned count_sum = 0;
  for (size_t i = 0; i < categories(); i++) {
    const unsigned ni = value.get<unsigned>(i);
    count_sum += ni;
    counts_[i] += ni;
    ratio_ -= fast_lgamma(ni + 1);
  }
  ratio_ += fast_lgamma(count_sum + 1);
}

void
dm_group::remove_value(const hypers &m, const value_accessor &value, rng_t &rng)
{
  MICROSCOPES_ASSERT(value.shape() == categories());
  unsigned count_sum = 0;
  for (size_t i = 0; i < categories(); i++) {
    const unsigned ni = value.get<unsigned>(i);
    count_sum += ni;
    MICROSCOPES_ASSERT(counts_[i] >= ni);
    counts_[i] -= ni;
    ratio_ += fast_lgamma(ni + 1);
  }
  ratio_ -= fast_lgamma(count_sum + 1);
}

float
dm_group::score_value(const hypers &m, const value_accessor &value, rng_t &rng) const
{
  const dm_hypers &h = static_cast<const dm_hypers &>(m);
  MICROSCOPES_ASSERT(categories() == h.categories());
  MICROSCOPES_ASSERT(value.shape() == categories());
  float score = 0.;
  unsigned x_sum = 0;
  float a_sum = 0.;
  unsigned n_sum = 0;
  for (size_t i = 0; i < categories(); i++) {
    const unsigned xi = value.get<unsigned>(i);
    const float ai = h.alphas()[i];
    const unsigned ni = counts_[i];
    x_sum += xi;
    a_sum += ai;
    n_sum += ni;
    score -= fast_lgamma(xi + 1);
    score += xi * fast_log(ai + ni);
  }
  score += fast_lgamma(x_sum + 1);
  score -= x_sum * fast_log(a_sum + n_sum);
  return score;
}

float
dm_group::score_data(const hypers &m, rng_t &rng) const
{
  const dm_hypers &h = static_cast<const dm_hypers &>(m);
  MICROSCOPES_ASSERT(categories() == h.categories());
  float score = ratio_;
  float alpha_sum = 0.;
  unsigned count_sum = 0;
  for (size_t i = 0; i < categories(); i++) {
    const float alpha = h.alphas()[i];
    const unsigned count = counts_[i];
    alpha_sum += alpha;
    count_sum += count;
    score += fast_lgamma(count + alpha)
           - fast_lgamma(alpha);
  }
  score += fast_lgamma(alpha_sum)
         - fast_lgamma(alpha_sum + count_sum);
  return score;
}

void
dm_group::sample_value(const hypers &m, value_mutator &value, rng_t &rng) const
{
  const dm_hypers &h = static_cast<const dm_hypers &>(m);
  (void)h;
  MICROSCOPES_ASSERT(categories() == h.categories());
  MICROSCOPES_ASSERT(value.shape() == categories());
  // XXX: we need a way to specify n, the # of samples from a categorical
  // distribution!
  //
  // sample_value() currently doesn't fit into our framework
  throw runtime_error("multinomial sampling unimplemented");
}

File: microscopes/dbg/models/dm.py
# Dirichlet-Multinomial conjugate prior, using the same framework as
# distributions

import numpy as np

from distributions.dbg.special import gammaln
from distributions.mixins import SharedMixin, GroupIoMixin, SharedIoMixin

NAME = 'DirichletMultinomial'
Value = int


class Shared(SharedMixin, SharedIoMixin):

    def __init__(self):
        self._alphas = None

    @property
    def dim(self):
        return len(self.alphas)

    def load(self, raw):
        self._alphas = np.array(raw['alphas'], dtype=np.float)

    def protobuf_load(self, message):
        self._alphas = np.array(message.alphas, dtype=np.float)

    def protobuf_dump(self, message):
        message.Clear()
        for alpha in self._alphas:
            message.alphas.append(alpha)


class Group(GroupIoMixin):

    def init(self, shared):
        self._counts = np.zeros(shared.dim, dtype=np.int)
        self._ratio = 0.

    def add_value(self, shared, value):
        count_sum = 0
        for i, xi in value:
            count_sum += xi
            self._counts[i] += xi
            self._ratio -= gammaln(xi + 1)
        self._ratio += gammaln(count_sum + 1)

    def remove_value(self, shared, value):
        count_sum = 0
        for i, xi in value:
            count_sum += xi
            self._counts[i] -= xi
            self._ratio += gammaln(xi + 1)
        self._ratio -= gammaln(count_sum + 1)

    def merge(self, shared, source):
        self._counts += source._counts
        self._ratio += source._ratio

    def score_value(self, shared, value):
        #x_sum = sum(value)
        #a_sum = sum(shared._alphas)
        #n_sum = sum(self._counts)
        #score = 0.
        #for xi, (ai, ni) in zip(value, zip(shared._alphas, self._counts)):
        #    score -= gammaln(xi + 1)
        #    score += xi * log(ai + ni)
        #score += gammaln(x_sum + 1)
        #score -= x_sum * log(a_sum + n_sum)
        #return score
        raise RuntimeError("need to fix")

    def score_data(self, shared):
        a_sum = sum(shared._alphas)
        n_sum = sum(self._counts)
        score = self._ratio
        for ai, ni in zip(self._alphas, self._counts):
            score += gammaln(ni + ai) - gammaln(ai)
        score += gammaln(a_sum) - gammaln(a_sum + n_sum)
        return score

    def sample_value(self, shared):
        raise RuntimeError("Unimplemented")

    def load(self, raw):
        self._counts = np.array(raw['counts'], dtype=np.int)
        self._ratio = float(raw['ratio'])

    def dump(self):
        return {'counts': self._counts.tolist(), 'ratio': self._ratio}

    def protobuf_load(self, message):
        self._counts = np.array(message.counts, dtype=np.int)
        self._ratio = message.ratio

    def protobuf_dump(self, message):
        message.Clear()
        for count in self._counts:
            message.counts.append(count)
        message.ratio = self._ratio
